tests/DataValues/TimeValueCalculatorTest.php

```diff
@@ -209,4 +209,450 @@ public function testGetSecondsForPrecision( $precision, $expected ) {
+	 */
+	private function provideTimestampsWithoutSign() {
+		return [
+			'0001-01-11T06:08:04Z',
+			'0016-02-11T06:08:04Z',
+			'0245-03-30T00:00:00Z',
+			'1054-04-11T14:00:02Z',
+			'2012-02-29T23:59:59Z',
+			'2012-03-01T00:00:00Z',
+			'2013-05-31T23:59:59Z',
+			'2014-06-01T00:00:00Z',
+			'54844518-07-25T02:00:00Z',
+			'5748404518-08-25T02:00:00Z',
+			'9990994999299999-09-09T14:20:00Z',
+			'9999999999999999-10-15T14:20:00Z',
+		];
+	}
+
+	/**
+	 * @return array of precisions from the most to the least precise
+	 */
+	private function provideAllPrecisions() {
+		return [
+			TimeValue::PRECISION_SECOND,
+			TimeValue::PRECISION_MINUTE,
+			TimeValue::PRECISION_HOUR,
+			TimeValue::PRECISION_DAY,
+			TimeValue::PRECISION_MONTH,
+			TimeValue::PRECISION_YEAR,
+			TimeValue::PRECISION_YEAR10,
+			TimeValue::PRECISION_YEAR100,
+			TimeValue::PRECISION_YEAR1K,
+			TimeValue::PRECISION_YEAR10K,
+			TimeValue::PRECISION_YEAR100K,
+			TimeValue::PRECISION_YEAR1M,
+			TimeValue::PRECISION_YEAR10M,
+			TimeValue::PRECISION_YEAR100M,
+			TimeValue::PRECISION_YEAR1G
+		];
+	}
+
+	/**
+	 * @return \Generator
+	 */
+	public function provideTimestamps() {
+		foreach ( [ '+', '-' ] as $sign ) {
+			foreach ( $this->provideTimestampsWithoutSign() as $timestamp ) {
+				yield $sign . $timestamp;
+			}
+		}
+	}
+
+	/**
+	 * @return \Generator
+	 */
+	public function provideTimeValuesHighVsLowPrecisions() {
+		$oldTimestamp = null;
+		$oldPrecision = null;
+		foreach ( $this->provideTimestamps() as $timestamp ) {
+			foreach ( $this->provideAllPrecisions() as $precision ) {
+				if ( !empty( $oldTimestamp ) && $timestamp === $oldTimestamp ) {
+					$oldTimeValue = new TimeValue(
+						$oldTimestamp,
+						0, 0, 0,
+						$oldPrecision,
+						TimeValue::CALENDAR_GREGORIAN
+					);
+					$timeValue = new TimeValue(
+						$timestamp,
+						0, 0, 0,
+						$precision,
+						TimeValue::CALENDAR_GREGORIAN
+					);
+					yield [ [ $oldTimeValue, $timeValue ] ];
+				}
+				$oldTimestamp = $timestamp;
+				$oldPrecision = $precision;
+			}
+		}
+	}
+
+	/**
+	 * @return \Generator
+	 */
+	public function provideTimestampsAndPrecisions() {
+		foreach ( $this->provideTimestamps() as $timestamp ) {
+			foreach ( $this->provideAllPrecisions() as $precision ) {
+				yield [ [ $timestamp, $precision ] ];
+			}
+		}
+	}
+
+	/**
+	 * @return \Generator of ISO 8601 timestamp arrays [ t1, t2 ] where t1 ≤ t2
+	 */
+	public function provideTimestampPairs() {
+		foreach ( [ '+', '-' ] as $sign ) {
+			$oldTimestamp = null;
+			foreach ( $this->provideTimestampsWithoutSign() as $timestamp ) {
+				$timestamp = $sign . $timestamp;
+				if ( $oldTimestamp !== null ) {
+					yield $sign === '+' ?
+						[ [ $oldTimestamp, $timestamp ] ] :
+						[ [ $timestamp, $oldTimestamp ] ];
+				}
+				$oldTimestamp = $timestamp;
+			}
+		}
+	}
+
+	/**
+	 * @return \Generator of arrays of:
+	 * - a TimeValue,
+	 * - its lower Unix timestamp, and
+	 * - its Unix higher timestamp.
```

Comment:
I feel like this should be consistent, either “lower/higher Unix timestamp” or “Unix lower/higher” timestamp, but not both :)

Reply:
Oh, right. xD